### DISCUSSION TITLE
undefined reference to `tmx::Map::Map()` | undefined reference to `vtable for tmx::ObjectGroup`

### DISCUSSION THREAD

**Discussion by durkisneer1** [fallahn/tmxlite#154] [Q&A] [answered]
I've used the sln file to build tmxlite into both debug and release modes, as well as tried the cmake compilation, and all three ways have been successful. The issue comes when attempting to use tmxlite in my own project. In both Visual Studio Code(with Makefile) and CLion(with CMake), I can't seem to get linking to work. Maybe I'm not compiling tmxlite correctly, or something isn't right in my CMake and Makefile configurations. Below I've pasted screenshots demonstrating the error in CLion, in VSC when building my library, and again in VSC when building my program, respectively. Thanks!

![image](https://github.com/fallahn/tmxlite/assets/103585220/9ade7d56-c8eb-4131-b0e2-8648a3c3b8a1)
![image](https://github.com/fallahn/tmxlite/assets/103585220/7f146fbc-d738-47d1-981d-654206f52414)
![image](https://github.com/fallahn/tmxlite/assets/103585220/e8e40b52-2e7b-4f65-9a46-1e190205f726)

**Comment by fallahn** [comment]
Hard to say as I've never really used either. Possibly:

- tmxlite needs to be built with the same compiler as your project - eg building with msvc then linking to a mingw project won't work
- linking order is wrong (though doesn't look like it) libs need to be linked in the order on which they depend on each other
- possibly the symbols aren't being exported correctly for mingw when the library is built

For the last point you could try building tmxlite as a static library, or check [here](https://github.com/fallahn/tmxlite/blob/master/tmxlite/include/tmxlite/Config.hpp) to make sure `TMXLITE_EXPORT_API` is being correctly defined for your compiler.

**Reply by durkisneer1** [reply]
Compiling in CLion using CMake with my mingw compiler rather than MSVC did indeed bring me further to a working project. But there's a new error coming up after linking. On a static build I get "undefined reference to `std::__throw_bad_array_new_length()'" upon compilation, and on a shared build I get the same message on a window popup upon launching the exe file. I keep seeing linux users only come across this error, which is strange since I'm on a Windows system. You got anything on this?

**Reply by fallahn** [reply, marked as answer]
I can't say I've experienced this personally. It sounds like a mismatch in libstdc++ linked when building the library vs building your project. EG tmxlite was built and linked to a version of libstdc++ which *does* have it (and then itself uses it) but your project is trying to link to a version of libstdc++ which *doesn't* have it - and you get the linker error. As libstdc++ is used by gcc/g++ it would probably occur more on linux than windows, simply because gcc is more prevalent there.

First I'd double check all your project settings to make sure you're using consistent linking between projects. Alternatively you *might* be able to hack around the problem by building with exceptions disabled (as this particular function is apparently designed as part of that mechinism) although I don't really recommend that as it's only masking the problem and it'll probably only lead to other unforseen complications 😔

**Reply by durkisneer1** [reply]
I finally discovered the problem. It was sort of what you and CGPT were saying about the compiler version, but also a little more than that. I was compiling tmxlite and my lib/exe to c++14, but the compiler I was using on tmxlite was 32-bit mingw instead of the 64-bit mingw I always use. I'm guessing that's where the "bad array length" came from.. Anyway, thanks for the help!
